sbi/inference/posteriors/mcmc_posterior.py

```diff
@@ -455,23 +464,40 @@ def sample_batched(
                 show_progress_bars=show_progress_bars,
             )
 
-        samples = self.theta_transform.inv(transformed_samples)
-        sample_shape_len = len(sample_shape)
-        # The MCMC sampler returns the samples per chain, of shape
-        # (num_samples, num_chains_extended, *input_shape). We return the samples as `
-        # (*sample_shape, x_batch_size, *input_shape). This means we want to combine
-        # all the chains that belong to the same x. However, using
-        # samples.reshape(*sample_shape,batch_size,-1) does not combine the samples in
-        #  the right order, since this mixes samples that belong to different `x`.
-        # This is a workaround to reshape the samples in the right order.
-        return samples.reshape((batch_size, *sample_shape, -1)).permute(  # type: ignore
-            tuple(range(1, sample_shape_len + 1))
+        # (num_chains_extended, samples_per_chain, *input_shape)
+        samples_per_chain: Tensor = self.theta_transform.inv(transformed_samples)  # type: ignore
+        input_shape = samples_per_chain.shape[-1]
+        # We need to collect samples for each x from the respective chains.
+        # However, using samples.reshape(*sample_shape,batch_size,-1) does not
+        #  combine the samples in the right order, since this mixes samples that
+        # belong to different `x`. The following is a workaround to reshape the
+        # samples in the right order.
+        samples_per_x = samples_per_chain.reshape((
+            batch_size,
+            # We leave the sample shape dim as -1 because we might have
+            # generated more samples than requested (more chains, or multiple of
+            # chains not matching sample_shape)
+            -1,
+            input_shape,
+        )).permute(
+            # NOTE: @Guy: I removed the `sample_shape_len` from the tuple below
+            # to allow putting -1 in the first position. Do you think the
+            # permute to sort the samples is still correct?
+            # tuple(range(1, sample_shape_len + 1))
+            tuple(range(1, 2))
```

Comment:
This is correct, but `tuple(range(1,2)) = (1,)`, so you can explicitly permute here with `.permute(1,0,-1)`. Since we've flattened `sample_shape` here (and reshape later), and `batch_shape` and `inpute_shape` are 1D here, we know that this tensor is 3D, so can simplify the code a bit.